### PR TITLE
Support platforms without strerror_r

### LINKED ISSUE #1064: strerror_r is not platform independent
See https://github.com/llvm/llvm-project/blob/3b984641a73780cac80c04610158c8167a8a786c/llvm/lib/Support/Errno.cpp for some code that works in windows.

cc @klausler

### PR DESCRIPTION
**Pull request by isuruf**
Fixed #1064

### CODE REVIEW

**Comment by isuruf** [comment]
Code taken from https://github.com/llvm/llvm-project/blob/3b984641a73780cac80c04610158c8167a8a786c/llvm/lib/Support/Errno.cpp

**Comment by CarolineConcatto** [comment]
Question, the use of llvm::sys::StrError instead of std::strerror solve that problem?

**Comment by isuruf** [comment]
@klausler mentioned that he didn't want LLVM libraries linked into the Fortran runtime which is why I'm copying that function instead of using it.

**Comment by DavidTruby** [comment]
Could you put a comment in the source referring to where this code comes from? I think we need a visible trail we can follow back in future!

File: runtime/io-error.cpp [hunk 1]
```diff
@@ -76,6 +77,32 @@ bool IoErrorHandler::GetIoMsg(char *buffer, std::size_t bufferLength) {
...
+#elif HAVE_DECL_STRERROR_S // "Windows Secure API"
+  return ::strerror_s(buffer, bufferLength, ioStat_) == 0;
+#elif HAVE_STRERROR
+  // Copy the thread un-safe result of strerror into
+  // the buffer as fast as possible to minimize impact
+  // of collision of strerror in multiple threads.
+  newBuf = strerror(ioStat_);
+#else
+  // Strange that this system doesn't even have strerror
+  return false;
```
Comment (sscalpone) [inline runtime/io-error.cpp]: Perhaps #error on systems without strerror would help people who are porting this code in the future?

File: runtime/io-error.cpp [hunk 2]
```diff
@@ -76,6 +77,32 @@ bool IoErrorHandler::GetIoMsg(char *buffer, std::size_t bufferLength) {
...
+#if HAVE_STRERROR_R
+  // strerror_r is thread-safe.
+#if defined(__GLIBC__) && defined(_GNU_SOURCE)
+  // glibc defines its own incompatible version of strerror_r
+  // which may not use the buffer supplied.
+  newBuf = ::strerror_r(ioStat_, buffer, bufferLength);
+#else
   return ::strerror_r(ioStat_, buffer, bufferLength) == 0;
+#endif
+#elif HAVE_DECL_STRERROR_S // "Windows Secure API"
```
Comment (sscalpone) [inline runtime/io-error.cpp]: Perhaps move this case first in case glibc ever defined strerror_s?

Comment (isuruf) [reply runtime/io-error.cpp]: I'd rather keep this code similar to the one in `llvm/lib/Support/Errno.cpp`

File: runtime/io-error.cpp (hunk 1 above)
Comment (isuruf) [reply runtime/io-error.cpp]: I've added an error in CMake code

**Comment by isuruf** [comment]
Squashed. Tests pass now

**Pull request merged**